Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,282 @@
+  /// should result in an ``RPCError`` with code ``RPCError/Code/failedPrecondition`` being thrown.
+  ///
+  /// If you want to forcefully cancel all active streams then cancel the task running ``connect(lazily:)``.
+  public func close() {
+    self.state.withLockedValue { state in
+      switch state {
+      case .unconnected:
+        state = .closed(0, nil)
+      case .connected(_, let openStreams, let signalEndConnection):
+        state = .closed(openStreams, signalEndConnection)
+        if openStreams == 0 {
+          signalEndConnection.finish()
+        }
+      case .closed:
+        ()
+      }
+    }
+  }
+
+  private enum WithStreamResult {
+    case success
+    case pending(AsyncStream<Void>)
+  }
+
+  /// Opens a stream using the transport, and uses it as input into a user-provided closure.
+  ///
+  /// - Important: The opened stream is closed after the closure is finished.
+  ///
+  /// Transport implementations should throw an ``RPCError`` with the following error codes:
+  /// - ``RPCError/Code/failedPrecondition`` if the transport is closing or has been closed.
+  /// - ``RPCError/Code/unavailable`` if it's temporarily not possible to create a stream and it
+  ///   may be possible after some backoff period.
+  ///
+  ///   This implementation will queue any streams (and thus block this call) if this function is called before
+  ///   ``connect(lazily:)``, until a connection is established - at which point all streams will be
+  ///   created.
+  ///
+  /// - Parameters:
+  ///   - descriptor: A description of the method to open a stream for.
+  ///   - closure: A closure that takes the opened stream as parameter.
+  /// - Returns: Whatever value was returned from `closure`.
+  public func withStream<T>(
+    descriptor: MethodDescriptor,
+    _ closure: (RPCStream<Inbound, Outbound>) async throws -> T
+  ) async throws -> T {
+    let request = RPCAsyncSequence<RPCRequestPart>.makeBackpressuredStream(watermarks: (16, 32))
+    let response = RPCAsyncSequence<RPCResponsePart>.makeBackpressuredStream(watermarks: (16, 32))
+
+    let clientStream = RPCStream(
+      descriptor: descriptor,
+      inbound: response.stream,
+      outbound: request.writer
+    )
+
+    let serverStream = RPCStream(
+      descriptor: descriptor,
+      inbound: request.stream,
+      outbound: response.writer
+    )
+
+    let result: Result<WithStreamResult, RPCError> = self.state.withLockedValue { state in
+      switch state {
+      case .connected(let serverTransport, let openStreams, let signalEndConnection):
+        do {
+          try serverTransport.acceptStream(serverStream)
+          state = .connected(serverTransport, openStreams + 1, signalEndConnection)
+        } catch let acceptStreamError as RPCError {
+          return .failure(acceptStreamError)
+        } catch {
+          return .failure(RPCError(code: .unknown, message: "Unknown error: \(error)."))
+        }
```

Comment:
(No need to change what you already have but an FYI in case you find it useful in the future)

We have an extension on `Result` to cast the error to a given type (or create one if the error is not of that type). I find it pretty handy when using result types.

```swift
return await Result { 
  ... 
}.castError(to: RPCError.self) {
  RPCError(...)
}
```